### ISSUE TITLE
static_frame non-determinism minimization

### ISSUE THREAD

**Issue by MeGaGiGaGon** [astral-sh/ty#2627]
### Summary

I noticed this library being non-deterministic in the primer runs, and decided to minimize it since it wasn't too big. Hopefully this is useful for tracking down the source of the non-determinisms. The non-deterministic error is:
```
error[invalid-return-type]: Return type does not match returned value
  --> static_frame\core\series.py:19:22
   |
18 | class Series(ContainerOperand, tp.Generic[TVIndex, TVDtype]):
19 |     def loc(self) -> InterGetItemLocReduces[TSeriesAny, TVDtype]:
   |                      ------------------------------------------- Expected `InterGetItemLocReduces[Series[Any, Any], TVDtype@Series]` because of return type
20 |         return InterGetItemLocReduces(self._extract_loc)
   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `InterGetItemLocReduces[Series[Any, Any], TVDtype@Series]`, found `InterGetItemLocReduces[Series[Any, Any] | Bottom[Index[Any]] | Any, TVDtype@Series]`
21 |     @tp.overload
22 |     def _extract_loc(self, key: TLocSelectorMany) -> tp.Self: ...
   |
info: rule `invalid-return-type` is enabled by default
```
I did the minimization using `while ($true) { ty check . -q }` and seeing if the count changed while deleting stuff. This is not 100% the smallest code possible, since leaving some of the supposedly useless imports made the non-det more common for me. I chose to stop here since it's still pretty small while still giving the non-det ~40% of the time. Your mileage will probably vary.

Some interesting things, it seems like some of the useless imports not only make the non-det more common, but actually are required. Some examples are the `from copy import deepcopy`s in `index.py` and `series.py`, as well as most of the useless imports in `node_selector.py`

For whatever reason the nesting, or maybe py.typed? also seems to be important, I only get the non-det running with the CWD as `static-frame-master`.

Zip is attached for convenience: [static-frame-master.zip](https://github.com/user-attachments/files/24854555/static-frame-master.zip)
Playground link, though the non-det doesn't seem to happen there: https://play.ty.dev/3b5cbe88-9bdd-4b47-8a40-fa6f0cc08333

`static-frame-master/static_frame/core/container.py`
```py
from __future__ import annotations


from static_frame.core.display import Display, DisplayActive
from static_frame.core.display_config import DisplayConfig, DisplayFormats
from static_frame.core.doc_str import doc_inject
from static_frame.core.exception import ErrorNotTruthy
from static_frame.core.interface_meta import InterfaceMeta

from static_frame.core.container_util import (
    a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,q,r,s,t,u,v,w,x,y,z  # More imports here makes the non-det more consistent
)


# -------------------------------------------------------------------------------
class ContainerBase(metaclass=InterfaceMeta):
    """
    Root of all containers. The core containers, like Series, Frame, and Index, inherit from ContainerOperand. The higher-order containers, like Bus, Quilt, Batch, and Yarn, inherit from ContainerBase.
    """

    __slots__ = ()

    def interface(self): ...


class ContainerOperandSequence(ContainerBase): ...


class ContainerOperand(ContainerOperandSequence): ...
```

`static-frame-master/static_frame/core/index.py`
```py
from __future__ import annotations

from copy import deepcopy

import typing_extensions as tp
from static_frame.core.container_util import (
    a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,q,r,s,t,u,v,w,x,y,z  # More imports here makes the non-det more consistent
)
from static_frame.core.container import ContainerOperandSequence


class IndexBase(ContainerOperandSequence): ...





TVDtype = tp.TypeVar('TVDtype')


class Index(IndexBase, tp.Generic[TVDtype]): ...
```

`static-frame-master/static_frame/core/node_selector.py`
```py
from __future__ import annotations

import numpy as np
import typing_extensions as tp
from numpy.ma import MaskedArray

from static_frame.core.exception import immutable_type_error_factory
from static_frame.core.util import (
    NULL_SLICE,
    TBlocKey,
    TCallableAny,
    TDepthLevelSpecifier,
    TDtypeSpecifier,
    TDtypesSpecifier,
    TILocSelector,
    TILocSelectorCompound,
    TILocSelectorMany,
    TILocSelectorOne,
    TLabel,
    TLocSelector,
    TLocSelectorMany,
)

# from static_frame.core.util import TCallableAny



if tp.TYPE_CHECKING:
    from static_frame.core.bus import Bus
    from static_frame.core.frame import (
        Frame,
        FrameAssignILoc,
        FrameAsType,
        FrameGO,
        FrameHE,
    )
    from static_frame.core.index import Index
    from static_frame.core.series import (
        Series
    )

# -------------------------------------------------------------------------------
TFrameOrSeries = tp.Union[
    'Frame[tp.Any, tp.Any, tp.Unpack[tp.Tuple[tp.Any, ...]]]',
    'Series[tp.Any, tp.Any]',
]

TVContainer_co = tp.TypeVar(
    'TVContainer_co',
    'Index[tp.Any]',
    'Series[tp.Any, tp.Any]',
    TFrameOrSeries,
    covariant=True,
)


TVDtype = tp.TypeVar('TVDtype', bound=np.generic, default=tp.Any)


class Interface:
    __slots__ = ()
    _INTERFACE: tp.Tuple[str, ...] = ()






class InterGetItemLocReduces(Interface, tp.Generic[TVContainer_co, TVDtype]):
    __slots__ = ('_func',)
    _INTERFACE = ('__getitem__',)

    _func: tp.Callable[[TLocSelector], TVContainer_co]

    def __init__(
        self,
        func: tp.Callable[[TLocSelector], TVContainer_co],
    ) -> None:
        self._func = func

    @tp.overload
    def __getitem__(self, key: TLocSelectorMany) -> TVContainer_co: ...

    @tp.overload
    def __getitem__(self, key: TLabel) -> TVDtype: ...

    def __getitem__(self, key: TLocSelector) -> tp.Any:
        return self._func(key)

    def __setitem__(self, key: TLabel, value: tp.Any) -> None:
        raise immutable_type_error_factory(
            self._func.__self__.__class__,  # type: ignore
            'loc',
            key,
            value,
        )
```

`static-frame-master/static_frame/core/series.py`
```py
from __future__ import annotations

from copy import deepcopy
import typing_extensions as tp

from static_frame.core.container import ContainerOperand
from static_frame.core.container_util import (
    a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,q,r,s,t,u,v,w,x,y,z  # More imports here makes the non-det more consistent
)
from static_frame.core.node_selector import InterGetItemLocReduces

TVDtype = tp.TypeVar('TVDtype')
TVIndex = tp.TypeVar('TVIndex')




class Series(ContainerOperand, tp.Generic[TVIndex, TVDtype]):
    def loc(self) -> InterGetItemLocReduces[TSeriesAny, TVDtype]:
        return InterGetItemLocReduces(self._extract_loc)
    @tp.overload
    def _extract_loc(self, key: TLocSelectorMany) -> tp.Self: ...
    @tp.overload
    def _extract_loc(self, key: TLabel) -> tp.Any: ...
    def _extract_loc(self, key: TLocSelector) -> tp.Any: ...

TSeriesAny = Series[tp.Any, tp.Any]
```

### Version

ty 0.0.13 (fc1478bd9 2026-01-21)

**Comment by carljm** [comment]
I'm not able to reproduce the non-determinism using the attached zip. I unzipped it, cd-ed into the `static-frame-master` directory, and ran a debug build of ty main (ruff commit b43a20408dafaba66e3f4cd3951f9d09c01f3938) on it, using `~/projects/ruff/target/debug/ty check . -q`. I get 17 diagnostics on every run; tried hundreds of runs.

This is on MacOS. @MeGaGiGaGon what OS did you repro on? I might try on a Linux VPS just to see if that makes a difference.

**Comment by carljm** [comment]
Even with `uvx ty@0.0.13 check . -q` (matching the version originally used in this repro) I always get 17 diagnostics.

**Comment by MeGaGiGaGon** [comment]
I made this on windows 10. The frequency looks to be OS dependent, since running on a Windows 11 machine I only see it ~1/20 runs. Also installing WSL currently to see how it performs on linux.

**Comment by MeGaGiGaGon** [comment]
Update: Running on WSL Ubuntu I only ever see 18 diagnostics, so it is definitely not working on other OSes

**Comment by carljm** [comment]
Going to close this as dupe of #1670, since I think we have a pretty good handle on the causes of non-determinism now, and several PRs in flight that will fix most cases. Thanks @MeGaGiGaGon
